File: js/spike/agents.ts
/**
 * SPIKE — Level-1/2 UX sugar over the substrate (see audit doc "The UX").
 * One new noun: agent(). Everything compiles to the six verbs in runtime.ts.
 *
 *   const explore = agent("explore", { does: "...", uses: { tools: [lookup] } })
 *   const coder   = agent("coder",   { does: "...", soul: "./AGENTS.md", team: [explore] })
 *   const r = await coder.run({ fetch: mockFetch }, "fix the bug")
 *
 *   const mia = agentFromDir("~/mia")          // the directory IS the agent
 *   const rt  = startRuntime({ fetch })         // Level 3 lives underneath
 */
import fs from "node:fs"
import path from "node:path"
import { defineTool, type Tool, type Request, type Answer } from "../dist/index.js"
import { AgentRuntime, type AgentDef, type Budget, type Handle, type RuntimeOptions, type TaskResult } from "./runtime.ts"

export interface AgentSpec {
  /** What the delegating model sees — the routing description. */
  does: string
  /** The toolkit VIEW for this agent (scoping = the whole security model). */
  uses?: { tools?: Tool[] }
  /** Identity: inline text, or a path to a soul file (AGENTS.md / SOUL.md). */
  soul?: string
  /** Path form of soul — read at build time. */
  soulFile?: string
  /** task-tool targets: listing agents here IS the subagent wiring. */
  team?: Agent[]
  budget?: Budget
  /** Model id; "inherit" = runtime default. */
  model?: string
  /** §10 interpreter authority for this agent's subtree. */
  waitFor?: (req: Request) => Promise<Answer>
  onSpawn?: AgentDef["onSpawn"]
  onClose?: AgentDef["onClose"]
}

export class Agent {
  readonly name: string
  readonly spec: AgentSpec
  constructor(name: string, spec: AgentSpec) {
    this.name = name
    this.spec = spec
  }

  /** Collect this agent + its whole team graph into a runtime registry. */
  registry(acc: Record<string, AgentDef> = {}): Record<string, AgentDef> {
    if (acc[this.name]) return acc
    let soul = this.spec.soul ?? ""
    if (this.spec.soulFile) soul = fs.readFileSync(this.spec.soulFile, "utf8")
    acc[this.name] = {
      name: this.name,
      description: this.spec.does,
      systemPrompt: soul,
      model: this.spec.model ?? "inherit",
      tools: this.spec.uses?.tools,
      budget: this.spec.budget,
      waitFor: this.spec.waitFor,
      onSpawn: this.spec.onSpawn,
      onClose: this.spec.onClose,
      /** team scoping: task targets = ONLY this agent's team (not the whole registry) */
      taskTargets: (this.spec.team ?? []).map((a) => a.name),
    } as AgentDef
    for (const t of this.spec.team ?? []) t.registry(acc)
    return acc
  }

  /** Level 1: one-shot — build a runtime, run to completion, tear down. */
  async run(rtOpts: Omit<RuntimeOptions, "registry">, prompt: string): Promise<TaskResult> {
    const rt = new AgentRuntime({ ...rtOpts, registry: this.registry() })
    const h = rt.spawn(rt.root, this.name)
    if ("isError" in h) return { text: h.isError, isError: true, status: "done", turns: 0, totalTokens: 0 }
    rt.wake(h, prompt)
    const r = await rt.wait(h)
    if (r.status !== "pending") await rt.close(rt.root)
    ;(r as TaskResult & { rt?: AgentRuntime }).rt = rt // exposed for durable resume
    return r
  }

  /** Bridge: an Agent IS a Tool — drop it into the OLD API's extraTools. */
  asTool(rtOpts: Omit<RuntimeOptions, "registry">): Tool {
    const self = this
    return defineTool({
      name: this.name,
      description: this.spec.does,
      inputSchema: { type: "object", properties: { prompt: { type: "string" } }, required: ["prompt"] },
      async run(args: Record<string, unknown>) {
        const r = await self.run(rtOpts, String(args.prompt))
        return { output: r.text, isError: r.isError, metadata: { agent: self.name, turns: r.turns, totalTokens: r.totalTokens } }
      },
    })
  }
}

/** The one new noun. */
export function agent(name: string, spec: AgentSpec): Agent {
  return new Agent(name, spec)
}

/**
 * Level 2: the directory IS the agent. Discovers (all optional, openclaw order):
 * AGENTS.md, SOUL.md, IDENTITY.md, USER.md, TOOLS.md, HEARTBEAT.md, MEMORY.md.
 * Injected as named sections at session start (cache doctrine: onSpawn reads memory in).
 */
export const BOOTSTRAP_ORDER = ["AGENTS.md", "SOUL.md", "IDENTITY.md", "USER.md", "TOOLS.md", "HEARTBEAT.md", "MEMORY.md"]
const MAX_BOOTSTRAP_FILE_BYTES = 2 * 1024 * 1024

export function agentFromDir(dir: string, opts?: Partial<AgentSpec> & { name?: string }): Agent {
  const sections: string[] = []
  for (const f of BOOTSTRAP_ORDER) {
    const p = path.join(dir, f)
    if (fs.existsSync(p)) {
      const stat = fs.statSync(p)
      let body = fs.readFileSync(p, "utf8")
      if (stat.size > MAX_BOOTSTRAP_FILE_BYTES) body = body.slice(0, MAX_BOOTSTRAP_FILE_BYTES) + "\n[truncated: file exceeds bootstrap cap]"
      sections.push(`## ${f}\n\n${body.trim()}`)
    }
  }
  const name = opts?.name ?? path.basename(dir)
  return agent(name, {
    does: opts?.does ?? `persona agent from ${dir}`,
    soul: sections.join("\n\n"),
    ...opts,
  })
}

/** The HEARTBEAT_OK sentinel — a silent wake produces no outbound message. */
export const HEARTBEAT_OK = "HEARTBEAT_OK"

export interface StartedAgent {
  handle: Handle
  rt: AgentRuntime
  stop: () => Promise<void>
  /** Fires only for NON-silent heartbeat results (HEARTBEAT_OK stripped). */
  onReport?: (text: string) => void
}

/**
 * Level 2: start a long-lived persona — heartbeat posts a tick to its own inbox
 * (unsolicited rail; ticks coalesce) and wakes it; HEARTBEAT_OK results stay silent.
 */
export function startAgent(
  a: Agent,
  rtOpts: Omit<RuntimeOptions, "registry">,
  opts?: { everyMs?: number; onReport?: (text: string) => void },
): StartedAgent {
  const rt = new AgentRuntime({ ...rtOpts, registry: a.registry() })
  const spawned = rt.spawn(rt.root, a.name)
  if ("isError" in spawned) throw new Error(spawned.isError)
  const handle = spawned
  let timer: ReturnType<typeof setInterval> | undefined
  if (opts?.everyMs) {
    timer = setInterval(() => {
      rt.post(handle, { from: "clock", channel: "timer", text: "tick" })
      if (handle.state === "idle") {
        void rt.runTurn(handle, "Heartbeat. Read your HEARTBEAT.md section and follow it. If nothing needs attention, reply HEARTBEAT_OK.").then((r) => {
          if (!r.isError && !r.text.includes(HEARTBEAT_OK)) opts.onReport?.(r.text)
        })
      }
    }, opts.everyMs)
  }
  return {
    handle,
    rt,
    stop: async () => {
      if (timer) clearInterval(timer)
      await rt.close(rt.root)
    },
  }
}

File: python/spike/agents.py
"""SPIKE — Level-1/2 UX sugar over the substrate (see audit doc "The UX").

One new noun: agent(). Everything compiles to the six verbs in runtime.py.

    explore = agent("explore", does="...", uses={"tools": [lookup]})
    coder   = agent("coder", does="...", soul_file="./AGENTS.md", team=[explore])
    r = await coder.run({"transport": mock}, "fix the bug")

    mia = agent_from_dir("~/mia")            # the directory IS the agent
    started = start_agent(mia, {...})        # Level 3 lives underneath

Mirrors js/spike/agents.ts.
"""
from __future__ import annotations

import asyncio
import os
from typing import Any, Callable, Optional

from toolnexus import Tool, ToolResult

from runtime import AgentDef, AgentRuntime, Budget, Handle, InboxItem, TaskResult

# Discovery order (all optional, openclaw order), injected as `## <file>` sections
# at session start (cache doctrine: on_spawn reads memory in).
BOOTSTRAP_ORDER = ["AGENTS.md", "SOUL.md", "IDENTITY.md", "USER.md", "TOOLS.md", "HEARTBEAT.md", "MEMORY.md"]
MAX_BOOTSTRAP_FILE_BYTES = 2 * 1024 * 1024

# The HEARTBEAT_OK sentinel — a silent wake produces no outbound message.
HEARTBEAT_OK = "HEARTBEAT_OK"


class Agent:
    def __init__(
        self,
        name: str,
        *,
        does: str,  # what the delegating model sees — the routing description
        uses: Optional[dict[str, Any]] = None,  # {"tools": [...]} — the toolkit VIEW (scoping)
        soul: Optional[str] = None,  # identity: inline text
        soul_file: Optional[str] = None,  # path form of soul — read at build time
        team: Optional[list["Agent"]] = None,  # task-tool targets: listing agents IS the wiring
        budget: Optional[Budget] = None,
        model: Optional[str] = None,  # "inherit" = runtime default
        wait_for: Optional[Callable[..., Any]] = None,  # §10 interpreter authority for the subtree
        on_spawn: Optional[Callable[..., Any]] = None,
        on_close: Optional[Callable[..., Any]] = None,
    ) -> None:
        self.name = name
        self.does = does
        self.uses = uses or {}
        self.soul = soul
        self.soul_file = soul_file
        self.team = team or []
        self.budget = budget
        self.model = model
        self.wait_for = wait_for
        self.on_spawn = on_spawn
        self.on_close = on_close

    def registry(self, acc: Optional[dict[str, AgentDef]] = None) -> dict[str, AgentDef]:
        """Collect this agent + its whole team graph into a runtime registry."""
        acc = acc if acc is not None else {}
        if self.name in acc:
            return acc
        soul = self.soul or ""
        if self.soul_file:
            with open(self.soul_file, encoding="utf-8") as f:
                soul = f.read()
        acc[self.name] = AgentDef(
            name=self.name,
            description=self.does,
            system_prompt=soul,
            model=self.model or "inherit",
            tools=self.uses.get("tools"),
            budget=self.budget,
            wait_for=self.wait_for,
            on_spawn=self.on_spawn,
            on_close=self.on_close,
            # team scoping: task targets = ONLY this agent's team (not the whole registry)
            task_targets=[a.name for a in self.team],
        )
        for t in self.team:
            t.registry(acc)
        return acc

    async def run(self, rt_opts: dict[str, Any], prompt: str) -> TaskResult:
        """Level 1: one-shot — build a runtime, run to completion, tear down."""
        rt = AgentRuntime(registry=self.registry(), **rt_opts)
        h = rt.spawn(rt.root, self.name)
        if not isinstance(h, Handle):
            return TaskResult(text=h.error, is_error=True, status="done")
        rt.wake(h, prompt)
        r = await rt.wait(h)
        if r.status != "pending":
            await rt.close(rt.root)
        r.rt = rt  # type: ignore[attr-defined] — exposed for durable resume
        return r

    def as_tool(self, rt_opts: dict[str, Any]) -> Tool:
        """Bridge: an Agent IS a Tool — drop it into the OLD API's extra_tools."""
        agent_self = self

        async def execute(args: Optional[dict[str, Any]] = None, ctx: Any = None) -> ToolResult:
            r = await agent_self.run(rt_opts, str((args or {}).get("prompt")))
            return ToolResult(
                output=r.text,
                is_error=r.is_error,
                metadata={"agent": agent_self.name, "turns": r.turns, "totalTokens": r.total_tokens},
            )

        return Tool(
            name=self.name,
            description=self.does,
            input_schema={"type": "object", "properties": {"prompt": {"type": "string"}}, "required": ["prompt"]},
            source="native",
            execute=execute,
        )


def agent(name: str, **spec: Any) -> Agent:
    """The one new noun."""
    return Agent(name, **spec)


def agent_from_dir(dir: str, *, name: Optional[str] = None, does: Optional[str] = None, **opts: Any) -> Agent:  # noqa: A002
    """Level 2: the directory IS the agent. Discovers the BOOTSTRAP_ORDER files
    (all optional) and injects them as named `## <file>` sections at session start."""
    sections: list[str] = []
    for f in BOOTSTRAP_ORDER:
        p = os.path.join(dir, f)
        if os.path.exists(p):
            size = os.stat(p).st_size
            with open(p, encoding="utf-8") as fh:
                body = fh.read()
            if size > MAX_BOOTSTRAP_FILE_BYTES:
                body = body[:MAX_BOOTSTRAP_FILE_BYTES] + "\n[truncated: file exceeds bootstrap cap]"
            sections.append(f"## {f}\n\n{body.strip()}")
    return Agent(
        name or os.path.basename(dir.rstrip("/")),
        does=does or f"persona agent from {dir}",
        soul="\n\n".join(sections),
        **opts,
    )


class StartedAgent:
    def __init__(self, handle: Handle, rt: AgentRuntime, stop: Callable[[], Any]) -> None:
        self.handle = handle
        self.rt = rt
        self._stop = stop

    async def stop(self) -> None:
        await self._stop()


def start_agent(
    a: Agent,
    rt_opts: dict[str, Any],
    *,
    every_ms: Optional[float] = None,
    on_report: Optional[Callable[[str], None]] = None,
) -> StartedAgent:
    """Level 2: start a long-lived persona — the heartbeat posts a tick to its own
    inbox (unsolicited rail; ticks coalesce) and wakes it; HEARTBEAT_OK results stay
    silent. NOTE (parity): JS uses setInterval fire-and-forget; here the heartbeat
    loop awaits each turn, so beats serialize instead of overlapping — the wake/
    coalesce semantics are identical, only the tick pacing under a slow turn differs.
    """
    rt = AgentRuntime(registry=a.registry(), **rt_opts)
    spawned = rt.spawn(rt.root, a.name)
    if not isinstance(spawned, Handle):
        raise RuntimeError(spawned.error)
    handle = spawned
    beat_task: Optional[asyncio.Task] = None
    if every_ms:

        async def beat() -> None:
            while True:
                await asyncio.sleep(every_ms / 1000.0)
                rt.post(handle, InboxItem(from_="clock", channel="timer", text="tick"))
                if handle.state == "idle":
                    r = await rt.run_turn(
                        handle,
                        "Heartbeat. Read your HEARTBEAT.md section and follow it. If nothing needs attention, reply HEARTBEAT_OK.",
                    )
                    if not r.is_error and HEARTBEAT_OK not in r.text and on_report is not None:
                        on_report(r.text)

        beat_task = asyncio.create_task(beat())

    async def stop() -> None:
        if beat_task is not None:
            beat_task.cancel()
            try:
                await beat_task
            except asyncio.CancelledError:
                pass
        await rt.close(rt.root)

    return StartedAgent(handle, rt, stop)
